{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#imports\n",
    "import pandas as pd\n",
    "import os\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metabat2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabatPath = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/results/metabat2/host_removed_checkv_filtered\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count how many bins for each sample.\n",
    "\n",
    "# Initialize an empty DataFrame to store binning statistics\n",
    "binning_df  = pd.DataFrame()\n",
    "\n",
    "# Iterate through directories in the 'metabatPath' directory\n",
    "for index,sample in enumerate(os.listdir(metabatPath)):\n",
    "    #print(sample)\n",
    "    # Initialize counters for the number of bins and unbinned contigs\n",
    "    num_bins = 0\n",
    "    num_unbinned = 0\n",
    "\n",
    "    # Create a DataFrame to store statistics for the current sample\n",
    "    sample_binning_df = pd.DataFrame()\n",
    "    \n",
    "    for bin in os.listdir(os.path.join(metabatPath, sample)):\n",
    "        \n",
    "         # Skip files starting with '.snakemake'\n",
    "        if bin.startswith('.snakemake'):\n",
    "            continue\n",
    "        #print(bin)\n",
    "        \n",
    "        #count num binned (it is only a bin if 2nd partion has a digit)\n",
    "        if bin.split(\".\")[1].isdigit():\n",
    "            num_bins += 1\n",
    "            #print(bin.split(\".\")[1])\n",
    "        #count many contigs were unbinned\n",
    "        if bin.split(\".\")[1] == \"unbinned\":            \n",
    "            with open(os.path.join(metabatPath,sample,bin), 'r') as fn:\n",
    "                for line in fn.readlines():\n",
    "                    if line.startswith(\">\"):\n",
    "                        num_unbinned += 1\n",
    "    #print(num_unbinned, sample)\n",
    "    #break\n",
    "\n",
    "    # Create a DataFrame for the current sample's statistics\n",
    "    sample_binning_df.at[index,'sample'] = sample\n",
    "    sample_binning_df.at[index,'num_bins'] = str(num_bins)\n",
    "    sample_binning_df.at[index,'num_unbinned_contigs'] = str(num_unbinned)\n",
    "\n",
    "\n",
    "    # Concatenate the sample DataFrame with the overall binning DataFrame\n",
    "    binning_df = pd.concat([binning_df, sample_binning_df])\n",
    "\n",
    "\n",
    "\n",
    "# Save the binning statistics DataFrame to a CSV file\n",
    "binning_df.to_csv(\"../supplimentary_data/metabat2_binning_stats.csv\", columns=binning_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>num_bins</th>\n",
       "      <th>num_unbinned_contigs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_64_AACGCATT-CAACCTGC_...</td>\n",
       "      <td>16</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_57_ATTATCAA-GAGAGTCG_...</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_44_TTAATCAG-CAGATTGG_...</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_53_TCTGCAAG-AAGGTGAA_...</td>\n",
       "      <td>24</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...</td>\n",
       "      <td>35</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_56_TTAATCAG-CAGATTGG_...</td>\n",
       "      <td>40</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_35_AGGTGCGA-GCCAGAAG_...</td>\n",
       "      <td>25</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_27_CTGTGGCG-GGCTAGTG_...</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_31_TCGCCTTG-GAATCAGC_...</td>\n",
       "      <td>26</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_71_TAAGCACA-TTAATGTC_...</td>\n",
       "      <td>20</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_12_GAACCTAG-AATACCAG_...</td>\n",
       "      <td>35</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_03_TAATGAAC-AAGCCAAG_...</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_81_GCCGTCGA-GTGATTAA_...</td>\n",
       "      <td>30</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_57_CTATGCGT-TGATATTC_...</td>\n",
       "      <td>22</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_06_CCAGTTAG-ATGACAGG_...</td>\n",
       "      <td>33</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_14_CGTTACCA-TAGCGTTC_...</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_63_TCGCTAGA-GCAGTACT_...</td>\n",
       "      <td>21</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               sample num_bins  \\\n",
       "0   HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...        6   \n",
       "1   HCFKNDSX3_104762-001-002_64_AACGCATT-CAACCTGC_...       16   \n",
       "2   HCFKNDSX3_104762-001-002_57_ATTATCAA-GAGAGTCG_...        9   \n",
       "3   HCFKNDSX3_104762-001-002_44_TTAATCAG-CAGATTGG_...       13   \n",
       "4   HCFKNDSX3_104762-001-002_53_TCTGCAAG-AAGGTGAA_...       24   \n",
       "5   HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...       35   \n",
       "6   H7NKYDSX3_104762-001-003_56_TTAATCAG-CAGATTGG_...       40   \n",
       "7   HCFKNDSX3_104762-001-002_35_AGGTGCGA-GCCAGAAG_...       25   \n",
       "8   HCFKNDSX3_104762-001-002_27_CTGTGGCG-GGCTAGTG_...       11   \n",
       "9   H7NKYDSX3_104762-001-003_31_TCGCCTTG-GAATCAGC_...       26   \n",
       "10  HCFKNDSX3_104762-001-002_71_TAAGCACA-TTAATGTC_...       20   \n",
       "11  HCFKNDSX3_104762-001-002_12_GAACCTAG-AATACCAG_...       35   \n",
       "12  HCFKNDSX3_104762-001-002_03_TAATGAAC-AAGCCAAG_...       15   \n",
       "13  H7NKYDSX3_104762-001-003_81_GCCGTCGA-GTGATTAA_...       30   \n",
       "14  H7NKYDSX3_104762-001-003_57_CTATGCGT-TGATATTC_...       22   \n",
       "15  H7NKYDSX3_104762-001-003_06_CCAGTTAG-ATGACAGG_...       33   \n",
       "16  HCFKNDSX3_104762-001-002_14_CGTTACCA-TAGCGTTC_...       15   \n",
       "17  HCFKNDSX3_104762-001-002_63_TCGCTAGA-GCAGTACT_...       21   \n",
       "\n",
       "   num_unbinned_contigs  \n",
       "0                     2  \n",
       "1                     7  \n",
       "2                     0  \n",
       "3                     1  \n",
       "4                     3  \n",
       "5                     8  \n",
       "6                     4  \n",
       "7                     4  \n",
       "8                     2  \n",
       "9                     6  \n",
       "10                    7  \n",
       "11                    6  \n",
       "12                    0  \n",
       "13                    2  \n",
       "14                    6  \n",
       "15                   10  \n",
       "16                    3  \n",
       "17                    2  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binning_df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VAMB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "host_exlusionPath =\"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/results/vamb/host_removed/bin_size1000/clusters.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in the cluster file\n",
    "def read_vamb_bins(filePath):\n",
    "    \"\"\"\n",
    "    Read VAMB bin information from a TSV file.\n",
    "\n",
    "    Parameters:\n",
    "    filePath (str): The path to the VAMB bin file in TSV format.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing bin information with columns 'bin_id' and 'contig_id'.\n",
    "    \"\"\"\n",
    "\n",
    "    vamb_df = pd.read_csv(filePath, sep=\"\\t\", header=None)\n",
    "\n",
    "    # Rename the columns for clarity\n",
    "    vamb_df.rename(columns={\n",
    "        0:'bin_id',\n",
    "        1:\"contig_id\"\n",
    "    },inplace=True)\n",
    "\n",
    "    return vamb_df\n",
    "\n",
    "# add sample ids to vamb bins\n",
    "def add_sample(x):\n",
    "    \"\"\"\n",
    "    Extract and add sample IDs to contig IDs.\n",
    "\n",
    "    Parameters:\n",
    "    x (str): A contig ID with a sample identifier.\n",
    "\n",
    "    Returns:\n",
    "    str: A modified contig ID with 'sample_' prefix.\n",
    "    \"\"\"\n",
    "    \n",
    "    sample_id = x.split(\"_\")[1]\n",
    "    \n",
    "    return f\"sample_{sample_id}\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#count the number of bins per sample\n",
    "def count_vamb_bin_per_sample(vamb_df):\n",
    "    \"\"\"\n",
    "    Count the number of bins per sample in a VAMB DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    vamb_df (pd.DataFrame): A DataFrame containing bin information with 'bin_id' and 'contig_id' columns.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame with the number of bins per sample, with columns 'sample' and 'num_bins'.\n",
    "    \"\"\"\n",
    "\n",
    "    bins_df = pd.DataFrame(vamb_df.groupby(['sample'])['bin_id'].count())\n",
    "    bins_df.reset_index(inplace=True)\n",
    "    bins_df.rename(columns={    \n",
    "        \"bin_id\":\"num_bins\"\n",
    "    }, inplace=True)\n",
    "\n",
    "    return bins_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>451582</th>\n",
       "      <td>351250</td>\n",
       "      <td>sample_8_NODE_5741_length_441_cov_9.831210</td>\n",
       "      <td>sample_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451583</th>\n",
       "      <td>351251</td>\n",
       "      <td>sample_18_NODE_3310_length_436_cov_8.132686</td>\n",
       "      <td>sample_18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451584</th>\n",
       "      <td>351252</td>\n",
       "      <td>sample_15_NODE_22368_length_308_cov_8.823204</td>\n",
       "      <td>sample_15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451585</th>\n",
       "      <td>351253</td>\n",
       "      <td>sample_14_NODE_773_length_1179_cov_13.676806</td>\n",
       "      <td>sample_14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451586</th>\n",
       "      <td>351254</td>\n",
       "      <td>sample_11_NODE_5313_length_557_cov_8.783721</td>\n",
       "      <td>sample_11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        bin_id                                     contig_id     sample\n",
       "451582  351250    sample_8_NODE_5741_length_441_cov_9.831210   sample_8\n",
       "451583  351251   sample_18_NODE_3310_length_436_cov_8.132686  sample_18\n",
       "451584  351252  sample_15_NODE_22368_length_308_cov_8.823204  sample_15\n",
       "451585  351253  sample_14_NODE_773_length_1179_cov_13.676806  sample_14\n",
       "451586  351254   sample_11_NODE_5313_length_557_cov_8.783721  sample_11"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read host exclusion bins\n",
    "host_exlusion_df = read_vamb_bins(host_exlusionPath)\n",
    "#add sample id\n",
    "host_exlusion_df['sample'] = host_exlusion_df['contig_id'].apply(add_sample)\n",
    "host_exlusion_df.tail()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>num_bins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sample_1</td>\n",
       "      <td>5471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sample_10</td>\n",
       "      <td>18851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sample_11</td>\n",
       "      <td>46995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sample_12</td>\n",
       "      <td>29873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sample_13</td>\n",
       "      <td>17474</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample  num_bins\n",
       "0   sample_1      5471\n",
       "1  sample_10     18851\n",
       "2  sample_11     46995\n",
       "3  sample_12     29873\n",
       "4  sample_13     17474"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "host_exl_bin = count_vamb_bin_per_sample(host_exlusion_df)\n",
    "host_exl_bin .head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "#same steps above for viruses-only\n",
    "viruses_onlyPath = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/results/vamb/virus_only/bin_size1000/clusters.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7708</th>\n",
       "      <td>5072</td>\n",
       "      <td>sample_4_NODE_279_length_282_cov_11.077419</td>\n",
       "      <td>sample_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7709</th>\n",
       "      <td>5073</td>\n",
       "      <td>sample_4_NODE_282_length_282_cov_0.361290</td>\n",
       "      <td>sample_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7710</th>\n",
       "      <td>5074</td>\n",
       "      <td>sample_14_NODE_304_length_341_cov_3.761682</td>\n",
       "      <td>sample_14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7711</th>\n",
       "      <td>5075</td>\n",
       "      <td>sample_4_NODE_62_length_513_cov_26.054404</td>\n",
       "      <td>sample_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7712</th>\n",
       "      <td>5076</td>\n",
       "      <td>sample_14_NODE_466_length_293_cov_10.198795</td>\n",
       "      <td>sample_14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      bin_id                                    contig_id     sample\n",
       "7708    5072   sample_4_NODE_279_length_282_cov_11.077419   sample_4\n",
       "7709    5073    sample_4_NODE_282_length_282_cov_0.361290   sample_4\n",
       "7710    5074   sample_14_NODE_304_length_341_cov_3.761682  sample_14\n",
       "7711    5075    sample_4_NODE_62_length_513_cov_26.054404   sample_4\n",
       "7712    5076  sample_14_NODE_466_length_293_cov_10.198795  sample_14"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read host exclusion bins\n",
    "viruses_only_df = read_vamb_bins(viruses_onlyPath)\n",
    "#add sample id\n",
    "viruses_only_df['sample'] = viruses_only_df['contig_id'].apply(add_sample)\n",
    "viruses_only_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>num_bins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sample_1</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sample_10</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sample_11</td>\n",
       "      <td>941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sample_12</td>\n",
       "      <td>924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sample_13</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sample_14</td>\n",
       "      <td>746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sample_15</td>\n",
       "      <td>413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sample_16</td>\n",
       "      <td>574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>sample_17</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sample_18</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>sample_2</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>sample_3</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>sample_4</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>sample_5</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>sample_6</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>sample_7</td>\n",
       "      <td>1603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>sample_8</td>\n",
       "      <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>sample_9</td>\n",
       "      <td>227</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample  num_bins\n",
       "0    sample_1        77\n",
       "1   sample_10       156\n",
       "2   sample_11       941\n",
       "3   sample_12       924\n",
       "4   sample_13       287\n",
       "5   sample_14       746\n",
       "6   sample_15       413\n",
       "7   sample_16       574\n",
       "8   sample_17        74\n",
       "9   sample_18        86\n",
       "10   sample_2       117\n",
       "11   sample_3       179\n",
       "12   sample_4       324\n",
       "13   sample_5       182\n",
       "14   sample_6       471\n",
       "15   sample_7      1603\n",
       "16   sample_8       332\n",
       "17   sample_9       227"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viruses_only_bin = count_vamb_bin_per_sample(viruses_only_df)\n",
    "viruses_only_bin.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>host_exclusion_bins</th>\n",
       "      <th>viruses-_only_bins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sample_1</td>\n",
       "      <td>5471</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sample_10</td>\n",
       "      <td>18851</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sample_11</td>\n",
       "      <td>46995</td>\n",
       "      <td>941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sample_12</td>\n",
       "      <td>29873</td>\n",
       "      <td>924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sample_13</td>\n",
       "      <td>17474</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sample_14</td>\n",
       "      <td>40316</td>\n",
       "      <td>746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sample_15</td>\n",
       "      <td>39804</td>\n",
       "      <td>413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sample_16</td>\n",
       "      <td>19624</td>\n",
       "      <td>574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>sample_17</td>\n",
       "      <td>10772</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sample_18</td>\n",
       "      <td>15964</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>sample_2</td>\n",
       "      <td>15683</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>sample_3</td>\n",
       "      <td>20009</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>sample_4</td>\n",
       "      <td>19874</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>sample_5</td>\n",
       "      <td>19302</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>sample_6</td>\n",
       "      <td>25638</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>sample_7</td>\n",
       "      <td>55211</td>\n",
       "      <td>1603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>sample_8</td>\n",
       "      <td>24967</td>\n",
       "      <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>sample_9</td>\n",
       "      <td>25759</td>\n",
       "      <td>227</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample  host_exclusion_bins  viruses-_only_bins\n",
       "0    sample_1                 5471                  77\n",
       "1   sample_10                18851                 156\n",
       "2   sample_11                46995                 941\n",
       "3   sample_12                29873                 924\n",
       "4   sample_13                17474                 287\n",
       "5   sample_14                40316                 746\n",
       "6   sample_15                39804                 413\n",
       "7   sample_16                19624                 574\n",
       "8   sample_17                10772                  74\n",
       "9   sample_18                15964                  86\n",
       "10   sample_2                15683                 117\n",
       "11   sample_3                20009                 179\n",
       "12   sample_4                19874                 324\n",
       "13   sample_5                19302                 182\n",
       "14   sample_6                25638                 471\n",
       "15   sample_7                55211                1603\n",
       "16   sample_8                24967                 332\n",
       "17   sample_9                25759                 227"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge the two counts\n",
    "merged_bins = pd.DataFrame(viruses_only_bin['sample'])\n",
    "merged_bins['host_exclusion_bins'] = host_exl_bin['num_bins']\n",
    "merged_bins['viruses-_only_bins'] = viruses_only_bin['num_bins']\n",
    "\n",
    "\n",
    "merged_bins.to_csv(\"../supplimentary_data/vamb_bins_per_sample.csv\", columns=merged_bins.columns)\n",
    "merged_bins.head(20)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
