{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#imports\n",
    "import pandas as pd\n",
    "import os\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import taxopy\n",
    "from ete3 import NCBITaxa\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare NCBI all Viral Nucleotide: \n",
    "\n",
    "prepare the 'lineage' file required by sourmash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_viral_metadataPath = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/tools/databases/AllNuclMetadata.csv.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3267704/1812190525.py:1: DtypeWarning: Columns (17) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ncbi_viral = pd.read_csv(ncbi_viral_metadataPath, usecols=[0,4,5,6,16,17]) #this file large select only useful cols\n"
     ]
    }
   ],
   "source": [
    "ncbi_viral = pd.read_csv(ncbi_viral_metadataPath, usecols=[0,4,5,6,16,17]) #this file large select only useful cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12516324, 6)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_viral.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Accession</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>Host</th>\n",
       "      <th>Isolation_Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_079174.1</td>\n",
       "      <td>Erwinia phage Fifi44</td>\n",
       "      <td>Loessnervirus</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Erwinia amylovora</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_079175.1</td>\n",
       "      <td>Escherichia phage PO103-1</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_079176.1</td>\n",
       "      <td>Escherichia phage MLP1</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Escherichia coli CFT073</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_079177.1</td>\n",
       "      <td>Yersinia phage PYps3T</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Yersinia pseudotuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_079178.1</td>\n",
       "      <td>Yersinia phage PYps23T</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Yersinia pseudotuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    #Accession                    Species              Genus        Family  \\\n",
       "0  NC_079174.1       Erwinia phage Fifi44      Loessnervirus  Chaseviridae   \n",
       "1  NC_079175.1  Escherichia phage PO103-1  Carltongylesvirus  Chaseviridae   \n",
       "2  NC_079176.1     Escherichia phage MLP1  Carltongylesvirus  Chaseviridae   \n",
       "3  NC_079177.1      Yersinia phage PYps3T  Carltongylesvirus  Chaseviridae   \n",
       "4  NC_079178.1     Yersinia phage PYps23T  Carltongylesvirus  Chaseviridae   \n",
       "\n",
       "                          Host Isolation_Source  \n",
       "0            Erwinia amylovora              NaN  \n",
       "1             Escherichia coli              NaN  \n",
       "2      Escherichia coli CFT073              NaN  \n",
       "3  Yersinia pseudotuberculosis              NaN  \n",
       "4  Yersinia pseudotuberculosis              NaN  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_viral.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ident</th>\n",
       "      <th>Superkingdom</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Host</th>\n",
       "      <th>Isolation_Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_079174.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Loessnervirus</td>\n",
       "      <td>Erwinia phage Fifi44</td>\n",
       "      <td>Erwinia amylovora</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_079175.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Escherichia phage PO103-1</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_079176.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Escherichia phage MLP1</td>\n",
       "      <td>Escherichia coli CFT073</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_079177.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Yersinia phage PYps3T</td>\n",
       "      <td>Yersinia pseudotuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_079178.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Yersinia phage PYps23T</td>\n",
       "      <td>Yersinia pseudotuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ident Superkingdom kingdom phylum order class        Family  \\\n",
       "0  NC_079174.1      viruses    None   None  None  None  Chaseviridae   \n",
       "1  NC_079175.1      viruses    None   None  None  None  Chaseviridae   \n",
       "2  NC_079176.1      viruses    None   None  None  None  Chaseviridae   \n",
       "3  NC_079177.1      viruses    None   None  None  None  Chaseviridae   \n",
       "4  NC_079178.1      viruses    None   None  None  None  Chaseviridae   \n",
       "\n",
       "               Genus                    Species                         Host  \\\n",
       "0      Loessnervirus       Erwinia phage Fifi44            Erwinia amylovora   \n",
       "1  Carltongylesvirus  Escherichia phage PO103-1             Escherichia coli   \n",
       "2  Carltongylesvirus     Escherichia phage MLP1      Escherichia coli CFT073   \n",
       "3  Carltongylesvirus      Yersinia phage PYps3T  Yersinia pseudotuberculosis   \n",
       "4  Carltongylesvirus     Yersinia phage PYps23T  Yersinia pseudotuberculosis   \n",
       "\n",
       "  Isolation_Source  \n",
       "0              NaN  \n",
       "1              NaN  \n",
       "2              NaN  \n",
       "3              NaN  \n",
       "4              NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#rename the Accession to ident as required by sourmash\n",
    "ncbi_viral.rename(columns={\"#Accession\":\"ident\"}, inplace=True)\n",
    "\n",
    "#add empty columns of other taxon leves;\n",
    "ncbi_viral[\"Superkingdom\"] = \"viruses\"\n",
    "ncbi_viral[\"kingdom\"] = None\n",
    "ncbi_viral[\"order\"] = None\n",
    "ncbi_viral[\"phylum\"] = None\n",
    "ncbi_viral[\"class\"] = None\n",
    "\n",
    "# reorder the cols\n",
    "ncbi_viral = ncbi_viral.iloc[:,[0,6,7,9,8,10,3,2,1,4,5]]\n",
    "ncbi_viral.head()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ident</th>\n",
       "      <th>superkingdom</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>Host</th>\n",
       "      <th>Isolation_Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_079174.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Loessnervirus</td>\n",
       "      <td>Erwinia phage Fifi44</td>\n",
       "      <td>Erwinia amylovora</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_079175.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Escherichia phage PO103-1</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_079176.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Escherichia phage MLP1</td>\n",
       "      <td>Escherichia coli CFT073</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_079177.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Yersinia phage PYps3T</td>\n",
       "      <td>Yersinia pseudotuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_079178.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Yersinia phage PYps23T</td>\n",
       "      <td>Yersinia pseudotuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ident superkingdom kingdom phylum order class        family  \\\n",
       "0  NC_079174.1      viruses    None   None  None  None  Chaseviridae   \n",
       "1  NC_079175.1      viruses    None   None  None  None  Chaseviridae   \n",
       "2  NC_079176.1      viruses    None   None  None  None  Chaseviridae   \n",
       "3  NC_079177.1      viruses    None   None  None  None  Chaseviridae   \n",
       "4  NC_079178.1      viruses    None   None  None  None  Chaseviridae   \n",
       "\n",
       "               genus                    species                         Host  \\\n",
       "0      Loessnervirus       Erwinia phage Fifi44            Erwinia amylovora   \n",
       "1  Carltongylesvirus  Escherichia phage PO103-1             Escherichia coli   \n",
       "2  Carltongylesvirus     Escherichia phage MLP1      Escherichia coli CFT073   \n",
       "3  Carltongylesvirus      Yersinia phage PYps3T  Yersinia pseudotuberculosis   \n",
       "4  Carltongylesvirus     Yersinia phage PYps23T  Yersinia pseudotuberculosis   \n",
       "\n",
       "  Isolation_Source  \n",
       "0              NaN  \n",
       "1              NaN  \n",
       "2              NaN  \n",
       "3              NaN  \n",
       "4              NaN  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#sourmash needs the col names in lower case ):\n",
    "ncbi_viral.rename(columns={\"Superkingdom\":\"superkingdom\",\n",
    "                           \"kingdom\":\"kingdom\",\n",
    "                           \"Family\":\"family\",\n",
    "                           \"Genus\":\"genus\",\n",
    "                           \"Species\":\"species\"}, inplace=True)\n",
    "ncbi_viral.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#write the selected columns to a gz compressed file \n",
    "csv = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/tools/databases/ncbi-all-viruses-2023-08-lineages.csv.gz\"\n",
    "ncbi_viral.to_csv(csv, columns=ncbi_viral.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3267704/674896567.py:2: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ncbi_viral = pd.read_csv(csv)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>ident</th>\n",
       "      <th>superkingdom</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>Host</th>\n",
       "      <th>Isolation_Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>NC_079174.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Loessnervirus</td>\n",
       "      <td>Erwinia phage Fifi44</td>\n",
       "      <td>Erwinia amylovora</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NC_079175.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Escherichia phage PO103-1</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>NC_079176.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Escherichia phage MLP1</td>\n",
       "      <td>Escherichia coli CFT073</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>NC_079177.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Yersinia phage PYps3T</td>\n",
       "      <td>Yersinia pseudotuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>NC_079178.1</td>\n",
       "      <td>viruses</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chaseviridae</td>\n",
       "      <td>Carltongylesvirus</td>\n",
       "      <td>Yersinia phage PYps23T</td>\n",
       "      <td>Yersinia pseudotuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0        ident superkingdom  kingdom  phylum  order  class  \\\n",
       "0           0  NC_079174.1      viruses      NaN     NaN    NaN    NaN   \n",
       "1           1  NC_079175.1      viruses      NaN     NaN    NaN    NaN   \n",
       "2           2  NC_079176.1      viruses      NaN     NaN    NaN    NaN   \n",
       "3           3  NC_079177.1      viruses      NaN     NaN    NaN    NaN   \n",
       "4           4  NC_079178.1      viruses      NaN     NaN    NaN    NaN   \n",
       "\n",
       "         family              genus                    species  \\\n",
       "0  Chaseviridae      Loessnervirus       Erwinia phage Fifi44   \n",
       "1  Chaseviridae  Carltongylesvirus  Escherichia phage PO103-1   \n",
       "2  Chaseviridae  Carltongylesvirus     Escherichia phage MLP1   \n",
       "3  Chaseviridae  Carltongylesvirus      Yersinia phage PYps3T   \n",
       "4  Chaseviridae  Carltongylesvirus     Yersinia phage PYps23T   \n",
       "\n",
       "                          Host Isolation_Source  \n",
       "0            Erwinia amylovora              NaN  \n",
       "1             Escherichia coli              NaN  \n",
       "2      Escherichia coli CFT073              NaN  \n",
       "3  Yersinia pseudotuberculosis              NaN  \n",
       "4  Yersinia pseudotuberculosis              NaN  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#loading the file (written out above; if all went well)\n",
    "csv = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/tools/databases/ncbi-all-viruses-2023-08-lineages.csv.gz\"\n",
    "ncbi_viral = pd.read_csv(csv)\n",
    "ncbi_viral.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get Taxid using ete3 package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ncbi = NCBITaxa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def process_sourmash_kreports(kreportPath):\n",
    "    \"\"\"\n",
    "    Process kreport files for multiple samples, extract relevant information, and consolidate it into a single DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    kreportPath (str): The directory containing kreport files for multiple samples.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing consolidated kreport data.\n",
    "    \"\"\"\n",
    "\n",
    "    kreport = pd.DataFrame()  # Initialize an empty DataFrame to store all samples' kreport data\n",
    "\n",
    "    # Initialize the NCBI Taxonomy database\n",
    "    ncbi = NCBITaxa()\n",
    "\n",
    "    # Iterate through files in the kreportPath\n",
    "    for samplefile in os.listdir(kreportPath):\n",
    "        if samplefile.endswith(\"k31.gather.genbank.kreport.txt\"):\n",
    "            sample_kreport = pd.read_csv(os.path.join(kreportPath, samplefile), sep=\"\\t\", header=None)\n",
    "\n",
    "            sample_kreport.rename(columns={\n",
    "                0: \"percent\",\n",
    "                1: \"Domain_BPs\",\n",
    "                2: \"species_BPs\",\n",
    "                3: \"taxonomy\",\n",
    "                4: \"taxid\",\n",
    "                5: \"name\"\n",
    "            }, inplace=True)\n",
    "\n",
    "            # Filter rows where 'taxonomy' is 'S' (indicating species)\n",
    "            sample_kreport = sample_kreport.query(\"taxonomy == 'S'\")\n",
    "\n",
    "            taxid_list = []\n",
    "\n",
    "            # Retrieve taxids for species names\n",
    "            for index, name in enumerate(sample_kreport['name']):\n",
    "                taxid_dict = ncbi.get_name_translator([name])\n",
    "                for key, id in taxid_dict.items():\n",
    "                    taxid_list.append(id[0])\n",
    "\n",
    "            sample_kreport['taxid'] = taxid_list\n",
    "            sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
    "\n",
    "            # Update the consolidated DataFrame with sample data\n",
    "            kreport = pd.concat([kreport, sample_kreport])\n",
    "\n",
    "    return kreport\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "kreportPath = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/results/sourmash/with-ncbi/scaled_300/output.sourmash-profiling/3-taxprofile\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n",
      "/tmp/ipykernel_3813430/4052086245.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['taxid'] = taxid_list\n",
      "/tmp/ipykernel_3813430/4052086245.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_kreport['sample'] = samplefile.split(\".\")[0]\n"
     ]
    }
   ],
   "source": [
    "kreport = process_sourmash_kreports(kreportPath)\n",
    "#kreport.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>percent</th>\n",
       "      <th>Domain_BPs</th>\n",
       "      <th>species_BPs</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>taxid</th>\n",
       "      <th>name</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>0.52</td>\n",
       "      <td>76999</td>\n",
       "      <td>76999</td>\n",
       "      <td>S</td>\n",
       "      <td>186617</td>\n",
       "      <td>uncultured marine virus</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>0.14</td>\n",
       "      <td>20000</td>\n",
       "      <td>20000</td>\n",
       "      <td>S</td>\n",
       "      <td>2202644</td>\n",
       "      <td>Microviridae sp.</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>0.12</td>\n",
       "      <td>17000</td>\n",
       "      <td>17000</td>\n",
       "      <td>S</td>\n",
       "      <td>2202563</td>\n",
       "      <td>CRESS virus sp.</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>0.07</td>\n",
       "      <td>10999</td>\n",
       "      <td>10999</td>\n",
       "      <td>S</td>\n",
       "      <td>1954248</td>\n",
       "      <td>Circoviridae sp.</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>0.05</td>\n",
       "      <td>8000</td>\n",
       "      <td>8000</td>\n",
       "      <td>S</td>\n",
       "      <td>43761</td>\n",
       "      <td>Barfin flounder nervous necrosis virus</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    percent  Domain_BPs  species_BPs taxonomy    taxid  \\\n",
       "47     0.52       76999        76999        S   186617   \n",
       "48     0.14       20000        20000        S  2202644   \n",
       "49     0.12       17000        17000        S  2202563   \n",
       "50     0.07       10999        10999        S  1954248   \n",
       "51     0.05        8000         8000        S    43761   \n",
       "\n",
       "                                      name  \\\n",
       "47                 uncultured marine virus   \n",
       "48                        Microviridae sp.   \n",
       "49                         CRESS virus sp.   \n",
       "50                        Circoviridae sp.   \n",
       "51  Barfin flounder nervous necrosis virus   \n",
       "\n",
       "                                               sample  \n",
       "47  HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...  \n",
       "48  HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...  \n",
       "49  HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...  \n",
       "50  HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...  \n",
       "51  HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kreport.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Link sourmash viruses to the host-db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#load host db viruses\n",
    "hostDbPath = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/scripts/virushostdb.tsv\"\n",
    "hostdb_viruses = pd.read_csv(hostDbPath, sep=\"\\t\", usecols=['virus tax id', \"refseq id\", \"virus lineage\", \"host name\", \"DISEASE\"])\n",
    "\n",
    "#select only viruses with human host\n",
    "hostdb_viruses.query(\" `host name` == 'Homo sapiens'\", inplace= True)\n",
    "human_viruses_taxids = [x for x in hostdb_viruses['virus tax id'] ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>percent</th>\n",
       "      <th>Domain_BPs</th>\n",
       "      <th>species_BPs</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>taxid</th>\n",
       "      <th>name</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>0.01</td>\n",
       "      <td>2000</td>\n",
       "      <td>2000</td>\n",
       "      <td>S</td>\n",
       "      <td>129951</td>\n",
       "      <td>Human mastadenovirus C</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>0.01</td>\n",
       "      <td>1000</td>\n",
       "      <td>1000</td>\n",
       "      <td>S</td>\n",
       "      <td>11983</td>\n",
       "      <td>Norwalk virus</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_71_TAAGCACA-TTAATGTC_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>0.03</td>\n",
       "      <td>1999</td>\n",
       "      <td>1999</td>\n",
       "      <td>S</td>\n",
       "      <td>129951</td>\n",
       "      <td>Human mastadenovirus C</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_35_AGGTGCGA-GCCAGAAG_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>0.02</td>\n",
       "      <td>2000</td>\n",
       "      <td>2000</td>\n",
       "      <td>S</td>\n",
       "      <td>129951</td>\n",
       "      <td>Human mastadenovirus C</td>\n",
       "      <td>H7NKYDSX3_104762-001-003_06_CCAGTTAG-ATGACAGG_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>0.01</td>\n",
       "      <td>2000</td>\n",
       "      <td>2000</td>\n",
       "      <td>S</td>\n",
       "      <td>11983</td>\n",
       "      <td>Norwalk virus</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_12_GAACCTAG-AATACCAG_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     percent  Domain_BPs  species_BPs taxonomy   taxid  \\\n",
       "61      0.01        2000         2000        S  129951   \n",
       "77      0.01        1000         1000        S   11983   \n",
       "63      0.03        1999         1999        S  129951   \n",
       "87      0.02        2000         2000        S  129951   \n",
       "103     0.01        2000         2000        S   11983   \n",
       "\n",
       "                       name                                             sample  \n",
       "61   Human mastadenovirus C  HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...  \n",
       "77            Norwalk virus  HCFKNDSX3_104762-001-002_71_TAAGCACA-TTAATGTC_...  \n",
       "63   Human mastadenovirus C  HCFKNDSX3_104762-001-002_35_AGGTGCGA-GCCAGAAG_...  \n",
       "87   Human mastadenovirus C  H7NKYDSX3_104762-001-003_06_CCAGTTAG-ATGACAGG_...  \n",
       "103           Norwalk virus  HCFKNDSX3_104762-001-002_12_GAACCTAG-AATACCAG_...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#\n",
    "human_v= kreport.query(\"taxid.isin(@human_viruses_taxids)\")\n",
    "human_v.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(848, 7)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kreport.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "sourmash_human_viruses = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/supplimentary_data/sourmash_k31_human_viruses.csv\"\n",
    "human_v.to_csv(sourmash_human_viruses, columns=human_v.columns)\n",
    "\n",
    "sourmash_all_viruses = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/supplimentary_data/sourmash_k31_all_viruses.csv\"\n",
    "kreport.to_csv(sourmash_all_viruses, columns=kreport.columns)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
