{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#imports\n",
    "import pandas as pd\n",
    "import os\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analyse Genomad Taxonomy assignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# iterate the sample files to read geNomad reports\n",
    "# genomadReportPath = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/results/genomad/host_removed\"\n",
    "# total_viruses = []\n",
    "# samples = []\n",
    "# genomad_taxonomy = pd.DataFrame()\n",
    "# for sample in os.listdir(genomadReportPath):\n",
    "#     if sample.startswith(\"all\"):\n",
    "#         continue\n",
    "#     ignore_dirs = [\"none_checkv_filtered\",\"genomad_and_blast_sample_viruse.csv\"] #files to ignore\n",
    "#     if sample in ignore_dirs:\n",
    "#         continue\n",
    "    \n",
    "#     #geNomad report per sample\n",
    "#     report_df = pd.read_csv(os.path.join(genomadReportPath, sample,\"checkv_medium_to_high_contigs_summary\",\n",
    "#                                          \"checkv_medium_to_high_contigs_virus_summary.tsv\"), sep=\"\\t\")\n",
    "    \n",
    "#     #count how many contigs had virus score above 0.7 (classified as virus)\n",
    "#     total_viruses.append(report_df.query(\"virus_score > 0.7\").shape[0])\n",
    "#     samples.append(sample)\n",
    "\n",
    "#     # select the geNomad taxonomies per sample\n",
    "#     report_df = report_df.query(\"virus_score > 0.7\") # take only those that viruses\n",
    "#     sample_genomad_taxonomy = pd.DataFrame({     \n",
    "#     'sample': sample,\n",
    "#     'contig_name': report_df['seq_name'],\n",
    "#     \"virus_score\": report_df['virus_score'],\n",
    "#     \"taxonomy\": report_df['taxonomy']\n",
    "#     }) \n",
    "\n",
    "#     #concat the taxonomies\n",
    "#     genomad_taxonomy = pd.concat([genomad_taxonomy, sample_genomad_taxonomy])\n",
    "\n",
    "\n",
    "# # count how many of the good contigs were viral (> 0.7 virus score)\n",
    "# viruses_df = pd.DataFrame({\n",
    "#     \"sample\": samples,\n",
    "#     \"genomad viruses\":total_viruses\n",
    "#         })\n",
    "\n",
    "\n",
    "# # write to file as supplimentary data   \n",
    "# #genomad_taxonomy.to_csv(\"../supplimentary_data/geNomad_taxonomies_for_good_quality_viral_contigs.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def process_genomad_reports(genomadReportPath):\n",
    "    \"\"\"\n",
    "    Process geNomad reports for viral contigs and generate summary statistics and taxonomies.\n",
    "\n",
    "    Parameters:\n",
    "    genomadReportPath (str): The directory containing geNomad reports for each sample.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing taxonomies for good quality viral contigs.\n",
    "    pd.DataFrame: A DataFrame containing the count of good quality viral contigs for each sample.\n",
    "    \"\"\"\n",
    "\n",
    "    total_viruses = []\n",
    "    samples = []\n",
    "    genomad_taxonomy = pd.DataFrame()\n",
    "\n",
    "    # Iterate through sample directories in the genomadReportPath\n",
    "    for sample in os.listdir(genomadReportPath):\n",
    "        if sample.startswith(\"all\"):\n",
    "            continue\n",
    "        ignore_dirs = [\"none_checkv_filtered\", \"genomad_and_blast_sample_viruse.csv\"]  # Files to ignore\n",
    "        if sample in ignore_dirs:\n",
    "            continue\n",
    "\n",
    "        # Read geNomad report per sample\n",
    "        report_df = pd.read_csv(os.path.join(genomadReportPath, sample, \"checkv_medium_to_high_contigs_summary\",\n",
    "                                            \"checkv_medium_to_high_contigs_virus_summary.tsv\"), sep=\"\\t\")\n",
    "\n",
    "        # Count how many contigs had a virus score above 0.7 (classified as virus)\n",
    "        total_viruses.append(report_df.query(\"virus_score > 0.7\").shape[0])\n",
    "        samples.append(sample)\n",
    "\n",
    "        # Select the geNomad taxonomies per sample\n",
    "        report_df = report_df.query(\"virus_score > 0.7\")  # Take only those classified as viruses\n",
    "        sample_genomad_taxonomy = pd.DataFrame({\n",
    "            'sample': sample,\n",
    "            'contig_name': report_df['seq_name'],\n",
    "            \"virus_score\": report_df['virus_score'],\n",
    "            \"taxonomy\": report_df['taxonomy']\n",
    "        })\n",
    "\n",
    "        # Concatenate the taxonomies\n",
    "        genomad_taxonomy = pd.concat([genomad_taxonomy, sample_genomad_taxonomy])\n",
    "\n",
    "    # Create a DataFrame with the count of good quality viral contigs for each sample\n",
    "    viruses_df = pd.DataFrame({\n",
    "        \"sample\": samples,\n",
    "        \"genomad viruses\": total_viruses\n",
    "    })\n",
    "\n",
    "    # Write taxonomies to a CSV file as supplementary data\n",
    "    genomad_taxonomy.to_csv(\"../supplimentary_data/geNomad_taxonomies_for_good_quality_viral_contigs.csv\")\n",
    "\n",
    "    return genomad_taxonomy, viruses_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>contig_name</th>\n",
       "      <th>virus_score</th>\n",
       "      <th>taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...</td>\n",
       "      <td>NODE_2_length_7715_cov_668.847654</td>\n",
       "      <td>0.9825</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...</td>\n",
       "      <td>NODE_10_length_2480_cov_46.993200</td>\n",
       "      <td>0.9712</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Kitrinoviricot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...</td>\n",
       "      <td>NODE_9_length_2491_cov_62.958968</td>\n",
       "      <td>0.9601</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Kitrinoviricot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...</td>\n",
       "      <td>NODE_16_length_2179_cov_20.280702</td>\n",
       "      <td>0.9572</td>\n",
       "      <td>Viruses;Monodnaviria;Sangervirae;Phixviricota;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...</td>\n",
       "      <td>NODE_3_length_4674_cov_144.778315</td>\n",
       "      <td>0.9398</td>\n",
       "      <td>Viruses;Riboviria</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              sample  \\\n",
       "0  HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...   \n",
       "1  HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...   \n",
       "2  HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...   \n",
       "3  HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...   \n",
       "4  HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...   \n",
       "\n",
       "                         contig_name  virus_score  \\\n",
       "0  NODE_2_length_7715_cov_668.847654       0.9825   \n",
       "1  NODE_10_length_2480_cov_46.993200       0.9712   \n",
       "2   NODE_9_length_2491_cov_62.958968       0.9601   \n",
       "3  NODE_16_length_2179_cov_20.280702       0.9572   \n",
       "4  NODE_3_length_4674_cov_144.778315       0.9398   \n",
       "\n",
       "                                            taxonomy  \n",
       "0  Viruses;Riboviria;Orthornavirae;Pisuviricota;P...  \n",
       "1  Viruses;Riboviria;Orthornavirae;Kitrinoviricot...  \n",
       "2  Viruses;Riboviria;Orthornavirae;Kitrinoviricot...  \n",
       "3  Viruses;Monodnaviria;Sangervirae;Phixviricota;...  \n",
       "4                                  Viruses;Riboviria  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomadReportPath = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/results/genomad/host_removed\"\n",
    "genomad_taxonomy, viruses_df = process_genomad_reports(genomadReportPath)\n",
    "genomad_taxonomy.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "460 contigs were classified at family level and 15 were unclassified\n"
     ]
    }
   ],
   "source": [
    "# examine how many contigs were classified at family level\n",
    "num_family = 0\n",
    "num_unclassified = 0\n",
    "\n",
    "for taxon in genomad_taxonomy['taxonomy']:\n",
    "    #count the number of unclassified\n",
    "    if taxon == 'Unclassified':\n",
    "        num_unclassified += 1\n",
    "        #print(taxon)\n",
    "        continue\n",
    "    #count number classified at family level\n",
    "    else:\n",
    "        levels = len(taxon.split(';'))\n",
    "        #print(levels)\n",
    "        if levels > 4: #https://portal.nersc.gov/genomad/taxonomic_assignment.html#\n",
    "            num_family +=1\n",
    " \n",
    "\n",
    "    #break\n",
    "print(f\"{num_family} contigs were classified at family level and {num_unclassified} were unclassified\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Viruses Associated with human host"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "hostDbPath = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/scripts/virushostdb.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>virus tax id</th>\n",
       "      <th>virus lineage</th>\n",
       "      <th>refseq id</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>host name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>46350</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>NC_001729</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>57579</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>NC_001829</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>57579</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>NC_001829</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>57579</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>NC_001829</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chlorocebus sabaeus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>68558</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>AF028704</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>10804</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>NC_001401</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>10804</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>NC_001401</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>68742</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>AF028705</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>82300</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>NC_006152</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>335103</td>\n",
       "      <td>Viruses; Riboviria; Orthornavirae; Duplornavir...</td>\n",
       "      <td>NC_007548, NC_007549, NC_007550, NC_007551, NC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   virus tax id                                      virus lineage  \\\n",
       "0         46350  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...   \n",
       "1         57579  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...   \n",
       "2         57579  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...   \n",
       "3         57579  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...   \n",
       "4         68558  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...   \n",
       "5         10804  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...   \n",
       "6         10804  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...   \n",
       "7         68742  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...   \n",
       "8         82300  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...   \n",
       "9        335103  Viruses; Riboviria; Orthornavirae; Duplornavir...   \n",
       "\n",
       "                                           refseq id DISEASE  \\\n",
       "0                                          NC_001729     NaN   \n",
       "1                                          NC_001829     NaN   \n",
       "2                                          NC_001829     NaN   \n",
       "3                                          NC_001829     NaN   \n",
       "4                                           AF028704     NaN   \n",
       "5                                          NC_001401     NaN   \n",
       "6                                          NC_001401     NaN   \n",
       "7                                           AF028705     NaN   \n",
       "8                                          NC_006152     NaN   \n",
       "9  NC_007548, NC_007549, NC_007550, NC_007551, NC...     NaN   \n",
       "\n",
       "             host name  \n",
       "0         Homo sapiens  \n",
       "1       Macaca mulatta  \n",
       "2         Homo sapiens  \n",
       "3  Chlorocebus sabaeus  \n",
       "4         Homo sapiens  \n",
       "5         Homo sapiens  \n",
       "6             Mammalia  \n",
       "7         Homo sapiens  \n",
       "8         Homo sapiens  \n",
       "9         Homo sapiens  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hostdb_viruses = pd.read_csv(hostDbPath, sep=\"\\t\", usecols=['virus tax id', \"refseq id\", \"virus lineage\", \"host name\", \"DISEASE\"])\n",
    "hostdb_viruses.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>virus tax id</th>\n",
       "      <th>virus lineage</th>\n",
       "      <th>refseq id</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>host name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>46350</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>NC_001729</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>57579</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>NC_001829</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>68558</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>AF028704</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>10804</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>NC_001401</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>68742</td>\n",
       "      <td>Viruses; Monodnaviria; Shotokuvirae; Cossaviri...</td>\n",
       "      <td>AF028705</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   virus tax id                                      virus lineage  refseq id  \\\n",
       "0         46350  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...  NC_001729   \n",
       "2         57579  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...  NC_001829   \n",
       "4         68558  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...   AF028704   \n",
       "5         10804  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...  NC_001401   \n",
       "7         68742  Viruses; Monodnaviria; Shotokuvirae; Cossaviri...   AF028705   \n",
       "\n",
       "  DISEASE     host name  \n",
       "0     NaN  Homo sapiens  \n",
       "2     NaN  Homo sapiens  \n",
       "4     NaN  Homo sapiens  \n",
       "5     NaN  Homo sapiens  \n",
       "7     NaN  Homo sapiens  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#select only viruses with human host\n",
    "hostdb_viruses.query(\" `host name` == 'Homo sapiens'\", inplace= True)\n",
    "hostdb_viruses.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig name</th>\n",
       "      <th>taxid</th>\n",
       "      <th>sacc</th>\n",
       "      <th>sscinames</th>\n",
       "      <th>sblastnames</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>pident</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NODE_1_length_10532_cov_285.994810</td>\n",
       "      <td>29159</td>\n",
       "      <td>XM_011415281</td>\n",
       "      <td>Crassostrea gigas</td>\n",
       "      <td>bivalves</td>\n",
       "      <td>1.500000e+00</td>\n",
       "      <td>52.8</td>\n",
       "      <td>100.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NODE_2_length_7700_cov_406.999868</td>\n",
       "      <td>11978</td>\n",
       "      <td>M86379</td>\n",
       "      <td>Feline calicivirus</td>\n",
       "      <td>viruses</td>\n",
       "      <td>8.200000e-07</td>\n",
       "      <td>73.1</td>\n",
       "      <td>100.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NODE_4_length_4017_cov_182.179177</td>\n",
       "      <td>1940570</td>\n",
       "      <td>MW348569</td>\n",
       "      <td>Parvoviridae sp.</td>\n",
       "      <td>viruses</td>\n",
       "      <td>1.600000e-155</td>\n",
       "      <td>566.0</td>\n",
       "      <td>74.943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NODE_5_length_3448_cov_97.686239</td>\n",
       "      <td>246410</td>\n",
       "      <td>XM_001240667</td>\n",
       "      <td>Coccidioides immitis RS</td>\n",
       "      <td>ascomycete fungi</td>\n",
       "      <td>4.800000e-01</td>\n",
       "      <td>52.8</td>\n",
       "      <td>100.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>NODE_7_length_3159_cov_29.293206</td>\n",
       "      <td>186617</td>\n",
       "      <td>JX904654</td>\n",
       "      <td>uncultured marine virus</td>\n",
       "      <td>viruses</td>\n",
       "      <td>1.980000e-14</td>\n",
       "      <td>97.1</td>\n",
       "      <td>84.694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>NODE_13_length_2673_cov_133.439120</td>\n",
       "      <td>29159</td>\n",
       "      <td>LR761639</td>\n",
       "      <td>Crassostrea gigas</td>\n",
       "      <td>bivalves</td>\n",
       "      <td>1.000000e-01</td>\n",
       "      <td>54.7</td>\n",
       "      <td>100.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>NODE_38_length_1701_cov_29.395807</td>\n",
       "      <td>1954248</td>\n",
       "      <td>MT208579</td>\n",
       "      <td>Circoviridae sp.</td>\n",
       "      <td>viruses</td>\n",
       "      <td>3.860000e-04</td>\n",
       "      <td>62.1</td>\n",
       "      <td>87.037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>NODE_52_length_1503_cov_82.534884</td>\n",
       "      <td>2202954</td>\n",
       "      <td>MH618202</td>\n",
       "      <td>Circular genetic element sp.</td>\n",
       "      <td>viruses</td>\n",
       "      <td>1.110000e-68</td>\n",
       "      <td>276.0</td>\n",
       "      <td>83.117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>NODE_61_length_1468_cov_163.879195</td>\n",
       "      <td>988011</td>\n",
       "      <td>OX411766</td>\n",
       "      <td>Orthosia incerta</td>\n",
       "      <td>moths</td>\n",
       "      <td>2.000000e-01</td>\n",
       "      <td>52.8</td>\n",
       "      <td>96.774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>NODE_100_length_1306_cov_35.586938</td>\n",
       "      <td>1954248</td>\n",
       "      <td>MH617385</td>\n",
       "      <td>Circoviridae sp.</td>\n",
       "      <td>viruses</td>\n",
       "      <td>4.900000e-02</td>\n",
       "      <td>54.7</td>\n",
       "      <td>100.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>NODE_133_length_1190_cov_20.523989</td>\n",
       "      <td>2202954</td>\n",
       "      <td>MH618185</td>\n",
       "      <td>Circular genetic element sp.</td>\n",
       "      <td>viruses</td>\n",
       "      <td>8.410000e-99</td>\n",
       "      <td>375.0</td>\n",
       "      <td>79.810</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           contig name    taxid          sacc  \\\n",
       "1   NODE_1_length_10532_cov_285.994810    29159  XM_011415281   \n",
       "11   NODE_2_length_7700_cov_406.999868    11978        M86379   \n",
       "28   NODE_4_length_4017_cov_182.179177  1940570      MW348569   \n",
       "29    NODE_5_length_3448_cov_97.686239   246410  XM_001240667   \n",
       "32    NODE_7_length_3159_cov_29.293206   186617      JX904654   \n",
       "34  NODE_13_length_2673_cov_133.439120    29159      LR761639   \n",
       "35   NODE_38_length_1701_cov_29.395807  1954248      MT208579   \n",
       "38   NODE_52_length_1503_cov_82.534884  2202954      MH618202   \n",
       "42  NODE_61_length_1468_cov_163.879195   988011      OX411766   \n",
       "45  NODE_100_length_1306_cov_35.586938  1954248      MH617385   \n",
       "53  NODE_133_length_1190_cov_20.523989  2202954      MH618185   \n",
       "\n",
       "                       sscinames       sblastnames         evalue  bitscore  \\\n",
       "1              Crassostrea gigas          bivalves   1.500000e+00      52.8   \n",
       "11            Feline calicivirus           viruses   8.200000e-07      73.1   \n",
       "28              Parvoviridae sp.           viruses  1.600000e-155     566.0   \n",
       "29       Coccidioides immitis RS  ascomycete fungi   4.800000e-01      52.8   \n",
       "32       uncultured marine virus           viruses   1.980000e-14      97.1   \n",
       "34             Crassostrea gigas          bivalves   1.000000e-01      54.7   \n",
       "35              Circoviridae sp.           viruses   3.860000e-04      62.1   \n",
       "38  Circular genetic element sp.           viruses   1.110000e-68     276.0   \n",
       "42              Orthosia incerta             moths   2.000000e-01      52.8   \n",
       "45              Circoviridae sp.           viruses   4.900000e-02      54.7   \n",
       "53  Circular genetic element sp.           viruses   8.410000e-99     375.0   \n",
       "\n",
       "     pident  \n",
       "1   100.000  \n",
       "11  100.000  \n",
       "28   74.943  \n",
       "29  100.000  \n",
       "32   84.694  \n",
       "34  100.000  \n",
       "35   87.037  \n",
       "38   83.117  \n",
       "42   96.774  \n",
       "45  100.000  \n",
       "53   79.810  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blastout = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/results/checkv/host_removed/HCFKNDSX3_104762-001-002_44_TTAATCAG-CAGATTGG_L004/checkv_medium_to_high_blast_test.txt\"\n",
    "detectedViruses = pd.read_csv(blastout,    sep=\"\\t\", header=None)\n",
    "detectedViruses.rename( columns={0:'contig name', 1:'taxid', 2: \"sacc\", 3: \"sscinames\", 4:\"sblastnames\", 5:\"evalue\", 6:\"bitscore\", 7:\"pident\"}, inplace=True)\n",
    "\n",
    "# keep only one hit per contig (query) based on eval, bit score and  percent identity\n",
    "detectedViruses.sort_values([\"contig name\", \"pident\", \"evalue\", \"bitscore\"], ascending= [True, False, True, False], inplace=True)\n",
    "detectedViruses.drop_duplicates(subset=['contig name'], keep='first', inplace=True)\n",
    "detectedViruses.sort_index(inplace=True)\n",
    "detectedViruses.head(20) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig name</th>\n",
       "      <th>taxid</th>\n",
       "      <th>sacc</th>\n",
       "      <th>sscinames</th>\n",
       "      <th>sblastnames</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>pident</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [contig name, taxid, sacc, sscinames, sblastnames, evalue, bitscore, pident]\n",
       "Index: []"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#select only viruses with human host\n",
    "hostdb_viruses.query(\" `host name` == 'Homo sapiens'\", inplace= True)\n",
    "human_viruses_taxids = [x for x in hostdb_viruses['virus tax id'] ]\n",
    "human_viruses_refseq1ds = [x for x in hostdb_viruses[\"refseq id\"] ] # MY BLAST DOESNT have refseqid; currently using subject accession number\n",
    "\n",
    "\n",
    " # select human viruses\n",
    "human_viruses_per_bin = detectedViruses.query(\"taxid.isin(@human_viruses_taxids) | \\\n",
    "                                                sacc.isin(@human_viruses_refseq1ds)\") #find common taxid or common refseq id. THERE DIFF BTWN refseq & ACCESSion NUM\n",
    "        \n",
    "\n",
    "human_viruses_per_bin"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a report of human viruses detected in our dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load host db viruses\n",
    "hostDbPath = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/scripts/virushostdb.tsv\"\n",
    "hostdb_viruses = pd.read_csv(hostDbPath, sep=\"\\t\", usecols=['virus tax id', \"refseq id\", \"virus lineage\", \"host name\", \"DISEASE\"])\n",
    "\n",
    "#select only viruses with human host\n",
    "hostdb_viruses.query(\" `host name` == 'Homo sapiens'\", inplace= True)\n",
    "human_viruses_taxids = [x for x in hostdb_viruses['virus tax id'] ]\n",
    "human_viruses_refseq1ds = [x for x in hostdb_viruses[\"refseq id\"] ] # MY BLAST DOESNT have refseqid; currently using subject accession number\n",
    "\n",
    "#Initiate empty dataframe to be filled with human viruses detected in our dataset and also one with all detected viruses\n",
    "all_detected_human_viruses = pd.DataFrame(columns=[ 'contig name', 'taxid', \"sacc\", \"sscinames\", \"sblastnames\", \"evalue\", \"bitscore\",\"pident\", \"sample\"])\n",
    "all_detected_viruses = pd.DataFrame(columns=[ 'contig name', 'taxid', \"sacc\", \"sscinames\", \"sblastnames\", \"evalue\", \"bitscore\",\"pident\", \"sample\"])\n",
    "\n",
    "\n",
    "detectedVirusesPath = \"/lustre/shared/wfsr-mcfa/projects/internships/luka/viral_metagenomics_pipeline/results/blast/host_removed\"\n",
    "samples = []\n",
    "num_viruses = []\n",
    "non_viral = []\n",
    "for file in os.listdir(detectedVirusesPath):\n",
    "    if file.endswith(\".txt\"):   \n",
    "        continue     \n",
    "    file_size = os.path.getsize(os.path.join(f'{detectedVirusesPath}',f'{file}', \"blastn_search.tsv\")) #some blast files are empty if there are no hits\n",
    "    \n",
    "    if  file_size == 0:\n",
    "        continue\n",
    "    \n",
    "    \n",
    "    detectedViruses = pd.read_csv(os.path.join(f'{detectedVirusesPath}',f'{file}', \"blastn_search.tsv\"), sep=\"\\t\", header=None)\n",
    "    detectedViruses.rename( columns={0:'contig name', 1:'taxid', 2: \"sacc\", 3: \"sscinames\", 4:\"sblastnames\", 5:\"evalue\", 6:\"bitscore\", 7:\"pident\"}, inplace=True)\n",
    "        \n",
    "    # keep only one hit per contig (query) based on eval, bit score and  percent identity: this could lead to missing other closely similar sequences\n",
    "    detectedViruses.sort_values([\"contig name\", \"bitscore\",\"pident\", \"evalue\"], ascending= [True, False, False, True], inplace=True)\n",
    "    detectedViruses.drop_duplicates(subset=['contig name'], keep='first', inplace=True)\n",
    "    detectedViruses.sort_index(inplace=True)\n",
    "        \n",
    "    #add a column with sample info\n",
    "    detectedViruses[\"sample\"] = file\n",
    "    #detectedViruses = detectedViruses.query(\"bitscore > 100\")\n",
    "    #merge all viruses\n",
    "    all_detected_viruses = pd.concat([all_detected_viruses, detectedViruses])\n",
    "\n",
    "    # select human viruses\n",
    "    human_viruses_per_sample = detectedViruses.query(\"taxid.isin(@human_viruses_taxids) | \\\n",
    "                                                    sacc.isin(@human_viruses_refseq1ds)\") #find common taxid or common refseq id. THERE DIFF BTWN refseq & ACCESSion NUM\n",
    "    \n",
    "    #check how many contigs were labeled as viruses by blast for comparison genomad\n",
    "    viruses_only = detectedViruses.query(\"sblastnames == 'viruses'\").shape[0]\n",
    "    num_viruses.append(viruses_only)\n",
    "    \n",
    "    \n",
    "    #check how many were not viruses\n",
    "    not_viruses = detectedViruses.query(\"sblastnames != 'viruses'\").shape[0]\n",
    "    non_viral.append(not_viruses)\n",
    "    \n",
    "    samples.append(file)\n",
    "    #print(file, detectedViruses.shape)\n",
    "    \n",
    "    #merge with human viruses for all bins\n",
    "    all_detected_human_viruses = pd.concat([all_detected_human_viruses, human_viruses_per_sample])\n",
    "\n",
    "#creat a df tracking the number of viruses per sample\n",
    "blast_num_viruses = pd.DataFrame({\n",
    "    \"sample\": samples,\n",
    "    \"blast viruses\": num_viruses,\n",
    "    \"blast_non_viral\": non_viral\n",
    "    })\n",
    "\n",
    "#merge blast viruses with genomad viruses\n",
    "genomad_blast_viruses = viruses_df.merge(blast_num_viruses, on=\"sample\")\n",
    "genomad_blast_viruses.to_csv(\"../supplimentary_data/genomad_and_blast_sample_viruses.csv\")\n",
    "# write out to file \n",
    "all_detected_human_viruses.to_csv(\"../supplimentary_data/good_quality_blast_host_removed_detected_human_viruses.txt\", sep='\\t', columns=all_detected_human_viruses.columns)\n",
    "all_detected_viruses.to_csv(\"../supplimentary_data/good_quality_blast_host_removed_all_detected_viruses.txt\", sep='\\t', columns=all_detected_human_viruses.columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>blast viruses</th>\n",
       "      <th>blast_non_viral</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_64_AACGCATT-CAACCTGC_...</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_57_ATTATCAA-GAGAGTCG_...</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_44_TTAATCAG-CAGATTGG_...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_53_TCTGCAAG-AAGGTGAA_...</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...</td>\n",
       "      <td>23</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_56_TTAATCAG-CAGATTGG_...</td>\n",
       "      <td>19</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_35_AGGTGCGA-GCCAGAAG_...</td>\n",
       "      <td>18</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_27_CTGTGGCG-GGCTAGTG_...</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_31_TCGCCTTG-GAATCAGC_...</td>\n",
       "      <td>16</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_71_TAAGCACA-TTAATGTC_...</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_12_GAACCTAG-AATACCAG_...</td>\n",
       "      <td>23</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_03_TAATGAAC-AAGCCAAG_...</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_81_GCCGTCGA-GTGATTAA_...</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_57_CTATGCGT-TGATATTC_...</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_06_CCAGTTAG-ATGACAGG_...</td>\n",
       "      <td>32</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_14_CGTTACCA-TAGCGTTC_...</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_63_TCGCTAGA-GCAGTACT_...</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               sample  blast viruses  \\\n",
       "0   HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...              3   \n",
       "1   HCFKNDSX3_104762-001-002_64_AACGCATT-CAACCTGC_...              8   \n",
       "2   HCFKNDSX3_104762-001-002_57_ATTATCAA-GAGAGTCG_...              4   \n",
       "3   HCFKNDSX3_104762-001-002_44_TTAATCAG-CAGATTGG_...              3   \n",
       "4   HCFKNDSX3_104762-001-002_53_TCTGCAAG-AAGGTGAA_...             18   \n",
       "5   HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...             23   \n",
       "6   H7NKYDSX3_104762-001-003_56_TTAATCAG-CAGATTGG_...             19   \n",
       "7   HCFKNDSX3_104762-001-002_35_AGGTGCGA-GCCAGAAG_...             18   \n",
       "8   HCFKNDSX3_104762-001-002_27_CTGTGGCG-GGCTAGTG_...              4   \n",
       "9   H7NKYDSX3_104762-001-003_31_TCGCCTTG-GAATCAGC_...             16   \n",
       "10  HCFKNDSX3_104762-001-002_71_TAAGCACA-TTAATGTC_...             10   \n",
       "11  HCFKNDSX3_104762-001-002_12_GAACCTAG-AATACCAG_...             23   \n",
       "12  HCFKNDSX3_104762-001-002_03_TAATGAAC-AAGCCAAG_...              8   \n",
       "13  H7NKYDSX3_104762-001-003_81_GCCGTCGA-GTGATTAA_...             14   \n",
       "14  H7NKYDSX3_104762-001-003_57_CTATGCGT-TGATATTC_...             13   \n",
       "15  H7NKYDSX3_104762-001-003_06_CCAGTTAG-ATGACAGG_...             32   \n",
       "16  HCFKNDSX3_104762-001-002_14_CGTTACCA-TAGCGTTC_...              9   \n",
       "17  HCFKNDSX3_104762-001-002_63_TCGCTAGA-GCAGTACT_...             10   \n",
       "\n",
       "    blast_non_viral  \n",
       "0                 1  \n",
       "1                 2  \n",
       "2                 2  \n",
       "3                 3  \n",
       "4                 1  \n",
       "5                 5  \n",
       "6                 4  \n",
       "7                 4  \n",
       "8                 0  \n",
       "9                 5  \n",
       "10                0  \n",
       "11                4  \n",
       "12                2  \n",
       "13                2  \n",
       "14                1  \n",
       "15                4  \n",
       "16                1  \n",
       "17                3  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_num_viruses.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig name</th>\n",
       "      <th>taxid</th>\n",
       "      <th>sacc</th>\n",
       "      <th>sscinames</th>\n",
       "      <th>sblastnames</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>pident</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NODE_33_length_1738_cov_11.656735</td>\n",
       "      <td>743300</td>\n",
       "      <td>NC_038392</td>\n",
       "      <td>Human stool-associated circular virus NG13</td>\n",
       "      <td>viruses</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>65.8</td>\n",
       "      <td>100.0</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_57_ATTATCAA-GAGAGTCG_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>NODE_164_length_1860_cov_96.292556</td>\n",
       "      <td>743300</td>\n",
       "      <td>NC_038392</td>\n",
       "      <td>Human stool-associated circular virus NG13</td>\n",
       "      <td>viruses</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>65.8</td>\n",
       "      <td>100.0</td>\n",
       "      <td>H7NKYDSX3_104762-001-003_31_TCGCCTTG-GAATCAGC_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           contig name   taxid       sacc  \\\n",
       "22   NODE_33_length_1738_cov_11.656735  743300  NC_038392   \n",
       "69  NODE_164_length_1860_cov_96.292556  743300  NC_038392   \n",
       "\n",
       "                                     sscinames sblastnames    evalue  \\\n",
       "22  Human stool-associated circular virus NG13     viruses  0.000030   \n",
       "69  Human stool-associated circular virus NG13     viruses  0.000033   \n",
       "\n",
       "    bitscore  pident                                             sample  \n",
       "22      65.8   100.0  HCFKNDSX3_104762-001-002_57_ATTATCAA-GAGAGTCG_...  \n",
       "69      65.8   100.0  H7NKYDSX3_104762-001-003_31_TCGCCTTG-GAATCAGC_...  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#all_detected_human_viruses.drop(columns=['subject_id'], inplace=True) #this col is not helpful i drop it\n",
    "all_detected_human_viruses.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(279, 9)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_detected_viruses.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### High confidence viruses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(204, 9)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "highconf_viruses = all_detected_viruses.query(\"bitscore > 100\")\n",
    "highconf_viruses.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig name</th>\n",
       "      <th>taxid</th>\n",
       "      <th>sacc</th>\n",
       "      <th>sscinames</th>\n",
       "      <th>sblastnames</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>pident</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NODE_2_length_7715_cov_668.847654</td>\n",
       "      <td>11978</td>\n",
       "      <td>M86379</td>\n",
       "      <td>Feline calicivirus</td>\n",
       "      <td>viruses</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>14022.0</td>\n",
       "      <td>99.622</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NODE_16_length_2179_cov_20.280702</td>\n",
       "      <td>2202644</td>\n",
       "      <td>MH572379</td>\n",
       "      <td>Microviridae sp.</td>\n",
       "      <td>viruses</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2715.0</td>\n",
       "      <td>98.882</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>NODE_64_length_1318_cov_46.818640</td>\n",
       "      <td>186617</td>\n",
       "      <td>JX904221</td>\n",
       "      <td>uncultured marine virus</td>\n",
       "      <td>viruses</td>\n",
       "      <td>1.680000e-46</td>\n",
       "      <td>202.0</td>\n",
       "      <td>73.623</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NODE_2_length_7705_cov_1225.754157</td>\n",
       "      <td>11978</td>\n",
       "      <td>M86379</td>\n",
       "      <td>Feline calicivirus</td>\n",
       "      <td>viruses</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>14020.0</td>\n",
       "      <td>99.622</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_64_AACGCATT-CAACCTGC_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NODE_11_length_4100_cov_167.561540</td>\n",
       "      <td>2202644</td>\n",
       "      <td>MH572286</td>\n",
       "      <td>Microviridae sp.</td>\n",
       "      <td>viruses</td>\n",
       "      <td>3.440000e-172</td>\n",
       "      <td>621.0</td>\n",
       "      <td>80.669</td>\n",
       "      <td>HCFKNDSX3_104762-001-002_64_AACGCATT-CAACCTGC_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           contig name    taxid      sacc  \\\n",
       "8    NODE_2_length_7715_cov_668.847654    11978    M86379   \n",
       "26   NODE_16_length_2179_cov_20.280702  2202644  MH572379   \n",
       "32   NODE_64_length_1318_cov_46.818640   186617  JX904221   \n",
       "0   NODE_2_length_7705_cov_1225.754157    11978    M86379   \n",
       "18  NODE_11_length_4100_cov_167.561540  2202644  MH572286   \n",
       "\n",
       "                  sscinames sblastnames         evalue  bitscore  pident  \\\n",
       "8        Feline calicivirus     viruses   0.000000e+00   14022.0  99.622   \n",
       "26         Microviridae sp.     viruses   0.000000e+00    2715.0  98.882   \n",
       "32  uncultured marine virus     viruses   1.680000e-46     202.0  73.623   \n",
       "0        Feline calicivirus     viruses   0.000000e+00   14020.0  99.622   \n",
       "18         Microviridae sp.     viruses  3.440000e-172     621.0  80.669   \n",
       "\n",
       "                                               sample  \n",
       "8   HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...  \n",
       "26  HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...  \n",
       "32  HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...  \n",
       "0   HCFKNDSX3_104762-001-002_64_AACGCATT-CAACCTGC_...  \n",
       "18  HCFKNDSX3_104762-001-002_64_AACGCATT-CAACCTGC_...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "highconf_viruses.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(204, 9)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "highconf_viruses.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>genomad viruses</th>\n",
       "      <th>blast viruses</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_64_AACGCATT-CAACCTGC_...</td>\n",
       "      <td>29</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_57_ATTATCAA-GAGAGTCG_...</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_44_TTAATCAG-CAGATTGG_...</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_53_TCTGCAAG-AAGGTGAA_...</td>\n",
       "      <td>33</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...</td>\n",
       "      <td>53</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_56_TTAATCAG-CAGATTGG_...</td>\n",
       "      <td>62</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_35_AGGTGCGA-GCCAGAAG_...</td>\n",
       "      <td>37</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_27_CTGTGGCG-GGCTAGTG_...</td>\n",
       "      <td>13</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_31_TCGCCTTG-GAATCAGC_...</td>\n",
       "      <td>48</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_71_TAAGCACA-TTAATGTC_...</td>\n",
       "      <td>38</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_12_GAACCTAG-AATACCAG_...</td>\n",
       "      <td>53</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_03_TAATGAAC-AAGCCAAG_...</td>\n",
       "      <td>21</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_81_GCCGTCGA-GTGATTAA_...</td>\n",
       "      <td>40</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_57_CTATGCGT-TGATATTC_...</td>\n",
       "      <td>36</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>H7NKYDSX3_104762-001-003_06_CCAGTTAG-ATGACAGG_...</td>\n",
       "      <td>70</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_14_CGTTACCA-TAGCGTTC_...</td>\n",
       "      <td>19</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HCFKNDSX3_104762-001-002_63_TCGCTAGA-GCAGTACT_...</td>\n",
       "      <td>29</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               sample  genomad viruses  \\\n",
       "0   HCFKNDSX3_104762-001-002_47_GAAGGAAG-CCAGGATG_...                8   \n",
       "1   HCFKNDSX3_104762-001-002_64_AACGCATT-CAACCTGC_...               29   \n",
       "2   HCFKNDSX3_104762-001-002_57_ATTATCAA-GAGAGTCG_...               10   \n",
       "3   HCFKNDSX3_104762-001-002_44_TTAATCAG-CAGATTGG_...               15   \n",
       "4   HCFKNDSX3_104762-001-002_53_TCTGCAAG-AAGGTGAA_...               33   \n",
       "5   HCFKNDSX3_104762-001-002_09_CATGATCG-GGAGAGTA_...               53   \n",
       "6   H7NKYDSX3_104762-001-003_56_TTAATCAG-CAGATTGG_...               62   \n",
       "7   HCFKNDSX3_104762-001-002_35_AGGTGCGA-GCCAGAAG_...               37   \n",
       "8   HCFKNDSX3_104762-001-002_27_CTGTGGCG-GGCTAGTG_...               13   \n",
       "9   H7NKYDSX3_104762-001-003_31_TCGCCTTG-GAATCAGC_...               48   \n",
       "10  HCFKNDSX3_104762-001-002_71_TAAGCACA-TTAATGTC_...               38   \n",
       "11  HCFKNDSX3_104762-001-002_12_GAACCTAG-AATACCAG_...               53   \n",
       "12  HCFKNDSX3_104762-001-002_03_TAATGAAC-AAGCCAAG_...               21   \n",
       "13  H7NKYDSX3_104762-001-003_81_GCCGTCGA-GTGATTAA_...               40   \n",
       "14  H7NKYDSX3_104762-001-003_57_CTATGCGT-TGATATTC_...               36   \n",
       "15  H7NKYDSX3_104762-001-003_06_CCAGTTAG-ATGACAGG_...               70   \n",
       "16  HCFKNDSX3_104762-001-002_14_CGTTACCA-TAGCGTTC_...               19   \n",
       "17  HCFKNDSX3_104762-001-002_63_TCGCTAGA-GCAGTACT_...               29   \n",
       "\n",
       "    blast viruses  \n",
       "0               3  \n",
       "1               8  \n",
       "2               4  \n",
       "3               3  \n",
       "4              18  \n",
       "5              23  \n",
       "6              19  \n",
       "7              18  \n",
       "8               4  \n",
       "9              16  \n",
       "10             10  \n",
       "11             23  \n",
       "12              8  \n",
       "13             14  \n",
       "14             13  \n",
       "15             32  \n",
       "16              9  \n",
       "17             10  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomad_blast_viruses"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequences labled as non-viral by BLAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0, 9)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_non_viruses = highconf_viruses.query(\"sblastnames != 'viruses'\")\n",
    "blast_non_viruses.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig name</th>\n",
       "      <th>taxid</th>\n",
       "      <th>sacc</th>\n",
       "      <th>sscinames</th>\n",
       "      <th>sblastnames</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>pident</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [contig name, taxid, sacc, sscinames, sblastnames, evalue, bitscore, pident, sample]\n",
       "Index: []"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_non_viruses.head(40)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
